controllers/helpers.go

```diff
@@ -597,3 +602,121 @@ func GetClusterIdentityFromRef(ctx context.Context, c client.Client, azureCluste
+				log.V(4).Info("Expected Cluster", "type", fmt.Sprintf("%T", e.Object))
+				return false
+			}
+			log = log.WithValues("namespace", c.Namespace, "cluster", c.Name)
+
+			// Only need to trigger a reconcile if the Cluster.Status.InfrastructureReady is true
+			if c.Spec.InfrastructureRef != nil {
+				log.V(6).Info("Cluster infrastructure is ready, allowing further processing")
+				return true
+			}
+
+			log.V(4).Info("Cluster infrastructure is not ready, blocking further processing")
+			return false
+		},
+		UpdateFunc:  func(e event.UpdateEvent) bool { return false },
+		DeleteFunc:  func(e event.DeleteEvent) bool { return false },
+		GenericFunc: func(e event.GenericEvent) bool { return false },
+	}
+}
+
+// ClusterUpdateInfraRefSet returns a predicate that returns true for an update event when a cluster has Status.InfrastructureReady changed from false to true
+// it also returns true if the resource provided is not a Cluster to allow for use with controller-runtime NewControllerManagedBy.
+func ClusterUpdateInfraRefSet(logger logr.Logger) predicate.Funcs {
+	return predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			log := logger.WithValues("predicate", "ClusterUpdateInfraRefSet", "eventType", "update")
+
+			oldCluster, ok := e.ObjectOld.(*clusterv1.Cluster)
+			if !ok {
+				log.V(4).Info("Expected Cluster", "type", fmt.Sprintf("%T", e.ObjectOld))
+				return false
+			}
+			log = log.WithValues("namespace", oldCluster.Namespace, "cluster", oldCluster.Name)
+
+			newCluster := e.ObjectNew.(*clusterv1.Cluster)
+
+			if oldCluster.Spec.InfrastructureRef == nil && newCluster.Spec.InfrastructureRef != nil {
+				log.V(6).Info("Cluster infrastructure became ready, allowing further processing")
+				return true
+			}
+
+			log.V(4).Info("Cluster infrastructure did not become ready, blocking further processing")
+			return false
+		},
+		CreateFunc:  func(e event.CreateEvent) bool { return false },
+		DeleteFunc:  func(e event.DeleteEvent) bool { return false },
+		GenericFunc: func(e event.GenericEvent) bool { return false },
+	}
+}
+
+// ClusterUnpausedAndInfraRefSet returns a Predicate that returns true on Cluster creation events where
+// both Cluster.Spec.Paused is false and Cluster.Status.InfrastructureReady is true and Update events when
+// either Cluster.Spec.Paused transitions to false or Cluster.Status.InfrastructureReady transitions to true.
+// This implements a common requirement for some cluster-api and provider controllers (such as Machine Infrastructure
+// controllers) to resume reconciliation when the Cluster is unpaused and when the infrastructure becomes ready.
+// Example use:
+//  err := controller.Watch(
+//      &source.Kind{Type: &clusterv1.Cluster{}},
+//      &handler.EnqueueRequestsFromMapFunc{
+//          ToRequests: clusterToMachines,
+//      },
+//      predicates.ClusterUnpausedAndInfraRefSet(r.Log),
+//  )
+func ClusterUnpausedAndInfraRefSet(logger logr.Logger) predicate.Funcs {
```

Comment:
@CecileRobertMichon I started adding a predicate for the update event. To be honest, it's a pretty complicated to add, and I'm wondering if it's worth it to refactor the controller code or just have the controller requeue instead as it seems like we're designing the predicate based around an edge case.